### ISSUE TITLE
Modifying an observable array does not update the view via the each directive

### ISSUE THREAD

**Issue by kiejo** [ryansolid/mobx-jsx#4]
Using `push` to add items to an observable array does not update the view using the `each` directive. Here is an example where clicking the "Add item" button does not add any new `li` to the rendered `ul`.

```jsx
import {r} from 'mobx-jsx'
import {observable} from 'mobx'

function App() {
  const items = observable([
    {name: 'Item 0'},
  ])

  const onAddItemClick = () => {
    items.push({name: 'Item ' + items.length})
  }

  return <MyList items={items} onAddClick={onAddItemClick} />
}

function MyList({items, onAddClick}) {
  return <>
    <ul>
      <$ each={items}>
        { item => <li>{(item.name + ` (items: ${items.length})`)}</li> }
      </$>
    </ul>
    <button onClick={onAddClick}>Add item</button>
  </>
}
```

**Comment by ryansolid** [comment]
Hmm I see.. It is interesting MobX allows this. See proxies can only detect access on properties. So by making the array itself observable only if the code itself accesses a property is a dependency made. However, the each control flow does not track key iterations for arrays since the performance would plummet (it would create N subscriptions instead of 1). It's much more expensive in large lists to manage the reactive graph than it is to just clone the Array. To illustrate, this logs hi once not twice:

```js
const a = observable([])
autorun(() => {
  const c = a;
  console.log('hi');
})
a.push('c')
```

And I suppose unsurprisingly(although it surprised me) this also only runs once:
```js
const a = mobx.observable({d: []})
mobx.autorun(() => {
  const c = a.d;
  console.log('hi');
})
a.d.push('c');
```
But this is where most reactive libraries differ. Ones based on Observables and likely older versions of MobX triggered the array any time a child was pushed removed, sorted etc.. So I understand why it acts this way to be consistent with Proxies but it is a bit unfortunate since without taking the hit of going into all the children can this change be tracked on a push. 

However if you treat arrays as primitive type, like a shallow immutable (it's rows can be mutated) it works. This is how I always treat arrays in reactive libraries so I hadn't hit this. Ie. this fires twice:
```js
const a = mobx.observable({d: []})
mobx.autorun(() => {
  const c = a.d;
  console.log('hi');
})
a.d = [...a.d, 'c']
```
However as you can imagine the array cannot be the top level observable for this to work with proxies. This does work with boxed observables without nesting.

I came to this pattern ages ago since I was getting sick of dealing with specialized array objects in these sort of libraries. Pre MobX 5 for instance an Observable Array wasn't actually an Array but an Array container. Rather than worry about the behavior and availability of special overridden built in methods I just started always treating them as frozen. I do appreciate Observable Arrays in MobX 5 are one of things they are pushing as Proxies finally enable them without the wrapper.

So tough spot. I'm pretty unwilling to change the the general plugin at this point to complete tank the performance across the board for all the libraries. But I do appreciate the ergonomic difficulties here. I realize now this is very opinionated perspective I've taken and not at all obvious. I'm hedging towards just documenting it. It is pretty common these days to handle arrays as frozen, both VDOM libraries like React do that, and even reactive libraries like Svelte.

**Comment by kiejo** [comment]
Thanks for looking into this in such detail.

> However, the each control flow does not track key iterations for arrays since the performance would plummet (it would create N subscriptions instead of 1). It's much more expensive in large lists to manage the reactive graph than it is to just clone the Array.

I definitely don't like the idea of creating N subscriptions, but I think it should be possible to get away with just a single subscription on the array level with a library like MobX. I'm pretty sure this is already what happens when you use methods like `slice`, `map`, `indexOf`, etc on an observable array in MobX. This is the code that handles these cases in MobX: https://github.com/mobxjs/mobx/blob/333dfe1719712f8e0729bca8b7f05960b2838f5c/src/types/observablearray.ts#L473

Adjusting my example code to access one of these methods (in this example `slice`) does indeed solve the issue and the view is automatically updated when `push` is used on the array:
```jsx
function MyList({items, onItemClick, onAddClick}) {
  function getItems() {
    items.slice() // Access slice to mark the array as observed
    return items
  }

  return <>
    <ul>
      <$ each={getItems()}>
        { item => <li model={item} onClick={onItemClick}>{(item.name)}</li> }
      </$>
    </ul>
    <button onClick={onAddClick}>Add item</button>
  </>
}
```
Observing arrays like this works out of the box in the context of React as there are no directives and the array methods can directly be used in the render method of a component. In React you would for example use `{items.map(item => <li>{(item.name)}</li>)}` instead of the directive, which will automatically mark the array as observed and rerender the component on array changes.

Wouldn't it be possible for this library to make use of the [observe](https://mobx.js.org/refguide/observe.html) feature of MobX when the `each` directive is used? This should also make it possible to clean up in case the directive gets "unmounted" as observe returns a dispose function.
On a side note, wouldn't it theoretically be possible to use the change data provided to the observe callback to skip most of the work done in `reconcileArrays`? I have no idea how much of a performance difference this would make, but it could be something interesting to explore.

So far I only have a rough understanding of how the different libraries (mobx-jsx, dom-expressions, babel-plugin-jsx-dom-expressions) work together, but I assume that such logic would ideally be implemented in mobx-jsx as it looks to be specific to how MobX works.

> It is pretty common these days to handle arrays as frozen, both VDOM libraries like React do that, and even reactive libraries like Svelte.

I agree that immutability seems to be the direction more and more libraries and frameworks are taking, but it is very opinionated and unfortunately makes it impossible to integrate with third-party libraries that need to mutate arrays or objects in place. I also like the idea of simply being able to use standard array methods and everything just working out of the box with no special rules. Obviously there are many trade-offs involved, but if a solution without a negative impact on performance is possible, it might be worth it.

**Comment by ryansolid** [comment]
Yeah all those methods basicallly iterate over the keys. Which is what is so cool about proxies, You don't need to write specialized methods and you still benefit on key access triggering the getters. So in those scenarios the naive approach would build up N deps. 

I actually started this with each library writing their own Each method, but found that my approach was faster than both Knockouts and MobX native versions. I tried to use MobX [createTransformer](https://mobx.js.org/refguide/create-transformer.html) to doing memoized optimized list updates etc and found that it was about 20% slower. I do believe in theory the fact that in fine grained you aren't losing the change information it should be able to spend less time diffing versus previous iteration (it would save memory too), however to date the cost of that granularity has been more expensive than the diffing. Believe it or not reconcileArray is faster than what MobX does. At the point they were all the same I decided that I could optimize even further bringing them into the runtime as I could reduce the whole process by a pass. For more details look at this issue: https://github.com/ryansolid/solid/issues/16 where someone is asking a similar question for one of my other libraries.

Observe hmm.. I've been using Autorun to date. It looks like it is additional hooks on top of the basic computed mechanism to do more with the value comparison stage. The problem I'm hitting right now in our example is simply the computation isn't firing at all since I'm not tracking the iteration. There might be something here as while writing a basic Reactive library is easy (can be done in like 90 lines of code) MobX has so many different methods and augments.

Ok all that being said I'm working on a game changer right now. I'm turning DOM Expressions from a runtime to a runtime generator. I was having an issue that using a Factory didn't allow for tree shaking and I want to be able to produce smaller bundles. This also means that I will be able to more easily compile different versions for MobX. Admittedly I started this project to prove a point. I like how it sits in front of Inferno and all React-Like libraries in performance, but for compatibility it's probably worth taking a hit. Especially with people starting to use it.

I suppose a shorter term solution is to see if I can just monkey patch the exported each method. But the way I've married the data update to the dom reconcilliation in a single pass means I'd essentially have to move a copy of reconcileArray into this library. It seems like a bit much. I'd like  to address this sooner than later but it isn't particularly simple. I'm hoping that over the next week or so as I'm working through the runtime generator an obvious answer comes to me.

**Comment by kiejo** [comment]
> So in those scenarios the naive approach would build up N deps.

That's what I initially thought too, but the comment in the MobX source code that I linked to makes it sound like there might be an optimization that circumvents this problem. Please note that I'm not really familiar with the source code and don't have a deep understanding of how Proxies work. I just had a quick look at the source code and thought it sounded promising. You might very well be right that it still builds up N deps. This is the comment for reference:
```
/**
 * Wrap function from prototype
 * Without this, everything works as well, but this works
 * faster as everything works on unproxied values
 */
```
> I actually started this with each library writing their own Each method, but found that my approach was faster than both Knockouts and MobX native versions.

Wow, that is really unexpected! Thanks for the details and the link to the other issue. This is really interesting and definitely not what I expected.

In general I think `autorun` is the correct method to use for this library as it automatically sets up observers based on the accessed properties. `observe` should be able to pick up changes on an array without the need to explicitly iterate over the keys, which I thought would make it a good fit for the `each` directive. I've used it before in different contexts and it worked quite well and I'm not sure what the problem in our example might be.

The runtime generator approach sounds really interesting and I can't wait to see what it will look like! I will make sure to follow the progress and there is no need from my side to come up with a short term solution for this particular issue. I'm playing around with this library as I really like the approach and seeing a performant and developer friendly alternative to the VDOM libraries. I will make sure to report issues that I run into during my testing as I would love to see how far this approach can be taken. Thanks for your great work and for sharing it!

**Comment by ryansolid** [comment]
Alright, so big changes to the runtime allows each framework to adopt their own syntax. I now have arrays working the way you expect (as well as support for class components). Instead of the old control flow, you use the `map` function that takes an observableArray as it's first argument and map iterator as the 2nd. It properly handles of MobX array mutators. I use the technique you highlighted in your earlier post so thanks for that. I still use my optimized memoized mapper as well so it's win-win.

Released in v0.6.0

**Comment by kiejo** [comment]
Thanks for the update, this sounds great!
